stdlib/runtime/exception.gr

```diff
@@ -0,0 +1,103 @@
+/* grainc-flags --compilation-mode=runtime */
+
+import WasmI32, {
+  eq as (==),
+  add as (+)
+} from "runtime/unsafe/wasmi32"
+
+import foreign wasm fd_write: (WasmI32, WasmI32, WasmI32, WasmI32) -> WasmI32 from "wasi_snapshot_preview1"
+
+enum Option<a> { Some(a), None }
+
+export let mut printers = 0n
+
+export let registerBasePrinter = (f) => {
+  let mut current = printers
+  while (true) {
+    // There will be at least one printer registered by the time this is called
+    let (_, next) = WasmI32.toGrain(current) : (Exception -> Option<String>, WasmI32)
+    if (next == 0n) {
+      // Using a tuple in runtime mode is typically disallowed as there is no way
+      // to reclaim the memory, but this function is only called once
+      let newBase = (WasmI32.fromGrain(f), 0n)
+      WasmI32.store(current, WasmI32.fromGrain(newBase), 12n)
+      break
+    }
+    current = next
+  }
+}
+
+export let dangerouslyRegisterPrinter = (f) => {
+  // This is dangerous because it leaks runtime memory.
+  // This should only be called by this module and Pervasives.
+  printers = WasmI32.fromGrain((f, printers))
+}
+
+let _GENERIC_EXCEPTION_NAME = "GrainException"
+
+let exceptionToString = (e: Exception) => {
+  let mut result = _GENERIC_EXCEPTION_NAME
+  let mut current = printers
+  while (true) {
+    if (current == 0n) {
+      break
+    }
+    let (printer, next) = WasmI32.toGrain(current) : (Exception -> Option<String>, WasmI32)
+    match (printer(e)) {
+      Some(str) => {
+        result = str
+        break
+      },
+      None => void
+    }
+    current = next
+  }
+  result
+}
+
+// HACK: Allocate static buffer for printing (40 bytes)
+// Would be nice to have a better way to allocate a static block from
+// the runtime heap, but this is the only module that needs to do it
```

Comment:
Do we want to create an issue for the nice-to-have?

Reply:
I'd want to make the issue if this ever comes up again.